### PR TITLE
Scn json translator

### PR DESCRIPTION
**Pull request by SHtress**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [x] Update documentation

### CODE REVIEW

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 1]
```diff
@@ -4,337 +4,633 @@
...
-  , mType(ST_NODE)
-  , mElementInfo(0)
+void uiSc2SCnJsonTranslator::runImpl()
 {
+  // get command arguments
+  sc_iterator5 * it5 = sc_iterator5_a_a_f_a_f_new(
+      s_default_ctx,
+      sc_type_node | sc_type_const,
+      sc_type_arc_common | sc_type_const,
+      mInputConstructionAddr,
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:60]:
```suggestion
      mStructureAddr,
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 2]
```diff
@@ -4,337 +4,633 @@
...
+  {
+    sc_iterator3 * it3 =
+        sc_iterator3_f_a_a_new(s_default_ctx, sc_iterator5_value(it5, 0), sc_type_arc_pos_const_perm, 0);
+    while (sc_iterator3_next(it3) == SC_TRUE)
+      mKeywordsList.insert(sc_iterator3_value(it3, 2));
+    sc_iterator3_free(it3);
+  }
+  sc_iterator5_free(it5);
+
+  collectScElementsInfo();
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:73]:
```suggestion
  collectStructureElementsInfo();
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 3]
```diff
@@ -4,337 +4,633 @@
...
+
+    // get begin/end addrs
+    sc_addr begAddr;
+    if (sc_memory_get_arc_begin(s_default_ctx, arcAddr, &begAddr) != SC_RESULT_OK)
+      continue;  // @todo process errors
+    sc_addr endAddr;
+    if (sc_memory_get_arc_end(s_default_ctx, arcAddr, &endAddr) != SC_RESULT_OK)
+      continue;  // @todo process errors
+
+    if (std::any_of(mFiltersList.cbegin(), mFiltersList.cend(), [begAddr](sc_addr const & modifier) {
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:104]:
```suggestion
    if (std::any_of(mFilterList.cbegin(), mFiltersList.cend(), [begAddr](sc_addr const & modifier) {
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 4]
```diff
@@ -4,337 +4,633 @@
...
+          return modifier == begAddr;
+        }))
     {
-      if (!(*it)->isInSentenceTree)
-        _createChildPredicate(*it);
+      filtered.insert(arcAddr);
+      filtered.insert(endAddr);
     }
+
+    sScElementInfo * elInfo = resolveElementInfo(it.first, it.second);
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:112]:
```suggestion
    ScStructureElementInfo * elInfo = resolveElementInfo(it.first, it.second);
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 5]
```diff
@@ -4,337 +4,633 @@
...
+    sScElementInfo * begInfo = resolveElementInfo(begAddr);
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:113]:
```suggestion
    sScElementInfo * begInfo = resolveStructureElementInfo(begAddr);
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 6]
```diff
@@ -4,337 +4,633 @@
...
+    sScElementInfo * endInfo = resolveElementInfo(endAddr);
+
+    elInfo->source = begInfo;
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:116]:
```suggestion
    elInfo->sourceInfo = begInfo;
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 7]
```diff
@@ -4,337 +4,633 @@
...
+    elInfo->target = endInfo;
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:117]:
```suggestion
    elInfo->targetInfo = endInfo;
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 8]
```diff
@@ -4,337 +4,633 @@
...
-    sScElementInfo * el = (mParent->mElementInfo == mElementInfo->target) ? mElementInfo->source : mElementInfo->target;
-    assert(el);
+    if (filtered.find(it.first) != filtered.cend())
+      continue;
 
-    uiSCnSentenceNode * child = new uiSCnSentenceNode();
-    child->mType = ST_NODE;
-    child->mElementInfo = el;
-    _appendChildNode(child);
+    sScElementInfo * edgeInfo = mScElementsInfo[it.first];
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:125]:
```suggestion
    sScElementInfo * edgeInfo = mStructureElementsInfo[it.first];
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 9]
```diff
@@ -4,337 +4,633 @@
...
+            }) != arc->inputArcs.cend())
+        {
+          keynodes.insert(arc->target);
+        }
+      });
+      // if there are no key elements, we do not collect structure elements
+      if (keynodes.empty())
+        continue;
+
+      sScElementInfo::tScElementInfoList removableArcs;
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:152]:
```suggestion
      sScElementInfo::StructureElementInfoList removableArcs;
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 10]
```diff
@@ -4,337 +4,633 @@
...
+          removableArcs.insert(elInfoOutputArc);
+        }
+      }
+      for (auto const & arc : removableArcs)
+      {
+        elInfo->outputArcs.erase(arc);
+      }
+
+      // get keyword from key elements
+      elInfo->structKeyword = findStructKeyword(keynodes);
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:174]:
```suggestion
      elInfo->structKeyword = findStructureKeyword(keynodes);
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 11]
```diff
@@ -4,337 +4,633 @@
...
+  });
+  if (result != elements.end())
+    return *result;
+
+  return nullptr;
+}
 
-  if (mType == ST_KEYWORD)
-    ss << ", \"keyword\": true";
+void uiSc2SCnJsonTranslator::json(sScElementInfo * elInfo, int level, bool isStruct, sc_json & result)
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:218]:
```suggestion
void uiSc2SCnJsonTranslator::parseScnJsonSentence(sScElementInfo * elInfo, int level, bool isStruct, sc_json & result)
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 12]
```diff
@@ -4,337 +4,633 @@
...
+{
+  bool isFullLinkedNodes = true;
 
-  if (mType == ST_PREDICATE)
+  getBaseInfo(elInfo, result);
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:222]:
```suggestion
  getElementInfo(elInfo, result);
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 13]
```diff
@@ -4,337 +4,633 @@
...
+  if (!elInfo->structureElements.empty())
   {
-    assert(mChildSentences.size() == 1);
-    ss << ", \"SCNode\" : " << (*mChildSentences.begin())->json();
+    structureElements = elInfo->structureElements;
+    json(elInfo->structKeyword, 0, true, result[scnTranslatorConstants::STRUCT.data()]);
   }
-  else
+  // if node is arc
+  if (elInfo->type & sc_type_arc_mask)
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:229]: Let's for each structure element type create alone method. Decompose this (json) method on atomic methods

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 14]
```diff
@@ -4,337 +4,633 @@
...
+      if (sc_helper_check_arc(s_default_ctx, elInfo->addr, format, sc_type_arc_common | sc_type_const) == SC_TRUE)
+      {
+        contentType = formatStr.data();
+        break;
+      }
+    }
+    // if the link has no format, then we get the content of the link
+    if (contentType.is_null())
+    {
+      sc_stream * stream;
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:254]: Content proccesing move to other method

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 15]
```diff
@@ -4,337 +4,633 @@
...
-}
-
-uiSCnSentenceNode * uiSCnSentenceNode::createChildNode(uiSCnSentenceNode::eSentenceNodeType type)
-{
-  uiSCnSentenceNode * child = new uiSCnSentenceNode();
-  child->mType = type;
-  _appendChildNode(child);
-  return child;
+  // if the nesting level is not greater than the maximum or the element is not a tuple, get children
+  if (level < maxLevel || (elInfo->type & sc_type_node && elInfo->type & sc_type_node_tuple))
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:277]:
```suggestion
  if (level < maxLevel || (elInfo->type & sc_type_node & sc_type_node_tuple))
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 16]
```diff
@@ -4,337 +4,633 @@
...
+    auto & resultChildren = result[scnTranslatorConstants::CHILDREN.data()];
+    // first get children from ordered list of modifiers
+    for (sc_addr modifier : mOrderList)
+    {
+      getChildrenByModifierAddr(elInfo, modifier, isStruct, resultChildren);
+    }
+    // then get the remaining children
+    getChildren(elInfo, isStruct, resultChildren);
+    if (isFullLinkedNodes)
+      getJsonOfLinkedNodes(resultChildren, level + 1, isStruct);
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:288]:
```suggestion
      parseLinkedNodesScnJson(resultChildren, level + 1, isStruct);
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 17]
```diff
@@ -4,337 +4,633 @@
...
-  _appendChildNode(child);
-  return child;
+  // if the nesting level is not greater than the maximum or the element is not a tuple, get children
+  if (level < maxLevel || (elInfo->type & sc_type_node && elInfo->type & sc_type_node_tuple))
+  {
+    auto & resultChildren = result[scnTranslatorConstants::CHILDREN.data()];
+    // first get children from ordered list of modifiers
+    for (sc_addr modifier : mOrderList)
+    {
+      getChildrenByModifierAddr(elInfo, modifier, isStruct, resultChildren);
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:283]:
```suggestion
      parseChildrenScnJsonByModifier(elInfo, modifier, isStruct, resultChildren);
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 18]
```diff
@@ -4,337 +4,633 @@
...
+    }
+    // then get the remaining children
+    getChildren(elInfo, isStruct, resultChildren);
+    if (isFullLinkedNodes)
+      getJsonOfLinkedNodes(resultChildren, level + 1, isStruct);
+  }
 }
 
-void uiSCnSentenceNode::destroyChilds()
+void uiSc2SCnJsonTranslator::getChildren(sScElementInfo * elInfo, bool isStruct, sc_json & children)
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:292]:
```suggestion
void uiSc2SCnJsonTranslator::parseChildrenScnJson(sScElementInfo * elInfo, bool isStruct, sc_json & children)
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 19]
```diff
@@ -4,337 +4,633 @@
...
 {
-  tSentenceNodeList::iterator it, itEnd = mChildSentences.end();
-  for (it = mChildSentences.begin(); it != itEnd; ++it)
-    delete *it;
-  for (it = mContourSentences.begin(); it != mContourSentences.end(); ++it)
-    delete *it;
+  getChildrenByDirection(elInfo->outputArcs, scnTranslatorConstants::RIGHT.data(), isStruct, children);
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:294]:
```suggestion
  parseChildrenScnJsonByDirection(elInfo->outputArcs, scnTranslatorConstants::RIGHT.data(), isStruct, children);
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 20]
```diff
@@ -4,337 +4,633 @@
...
+            modifierAddr == modifiers[0][scnTranslatorConstants::ADDR.data()]);
+      });
+      // if such element is found then merge children
+      if (result == children.end())
+      {
+        children.push_back(child);
+      }
+      else
+      {
+        (*result)[scnTranslatorConstants::ARCS.data()].push_back(child[scnTranslatorConstants::ARCS.data()][0]);
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:334]:
```suggestion
        (*result)[ScnJsonTranslatorConstants::ARCS.data()].push_back(child[scnTranslatorConstants::ARCS.data()][0]);
```

File: sc-memory/sc-core/sc-store/sc-fs-storage/sc_dictionary_fs_storage.c [hunk 21]
```diff
@@ -204,7 +204,8 @@ sc_bool sc_dictionary_fs_storage_get_sc_links_by_content(const sc_char * sc_stri
 
 void _sc_dictionary_fs_storage_update_links_list(sc_dictionary_node * node, void ** args)
 {
-  if (node->data_list != null_ptr)
+  sc_list * full_list = args[0];
+  if (node->data_list != null_ptr && full_list->size < 100)
```
Comment (NikitaZotov) [inline sc-memory/sc-core/sc-store/sc-fs-storage/sc_dictionary_fs_storage.c:208]: Remove it

File: sc-memory/sc-core/sc-store/sc-fs-storage/sc_dictionary_fs_storage.c [hunk 22]
```diff
@@ -261,7 +260,8 @@ sc_bool sc_dictionary_fs_storage_get_sc_links_by_content_substr(
 
 void _sc_dictionary_fs_storage_update_strings_list(sc_dictionary_node * node, void ** args)
 {
-  if (node->data_list != null_ptr)
+  sc_list * full_list = args[0];
+  if (node->data_list != null_ptr && full_list->size < 100)
```
Comment (NikitaZotov) [inline sc-memory/sc-core/sc-store/sc-fs-storage/sc_dictionary_fs_storage.c:264]: Remove it

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 23]
```diff
@@ -4,337 +4,633 @@
...
 #include "uiKeynodes.h"
 #include "uiUtils.h"
+#include <algorithm>
+#include <string_view>
 
-uiSCnSentenceNode::uiSCnSentenceNode(sScElementInfo * keywordEl)
-  : mParent(0)
-  , mType(ST_KEYWORD)
-  , mElementInfo(keywordEl)
+namespace scnTranslatorConstants
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:18]: Let's name all type and methods names with `CamelCase`

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

**Review by FallenChromium — APPROVED** [review]
The linker fix looks good to me

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.h [hunk 24]
```diff
@@ -8,91 +8,22 @@
...
-  //! List of child sentences
-  tSentenceNodeList mChildSentences;
-  //! Pointer to sc-element info
-  sScElementInfo * mElementInfo;
-  //! String that contains cache of generated json for tree
-  String mJSONData;
+  ScStructureElementInfo * sourceInfo;
+  ScStructureElementInfo * targetInfo;
+  ScStructureElementInfo * structKeyword;
+  StructureElementInfoList outputArcs;
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.h:22]:
```suggestion
  ScStructureElementInfoList outputArcs;
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.h [hunk 25]
```diff
@@ -109,31 +40,84 @@ class uiSc2SCnJsonTranslator : public uiTranslateFromSc
...
+
+  //! Get children by direction for specified arcs list
+  void ParseChildrenScnJsonByDirection(
+      ScStructureElementInfo::StructureElementInfoList const & arcs,
+      String const & direction,
+      bool isStruct,
+      sc_json & children);
+
+  //! Get base json information about specified element
+  void GetElementInfo(ScStructureElementInfo * elInfo, sc_json & result);
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.h:68]:
```suggestion
  void ParseScElementInfo(ScStructureElementInfo * elInfo, sc_json & result);
```

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.h [hunk 26]
```diff
@@ -109,31 +40,84 @@ class uiSc2SCnJsonTranslator : public uiTranslateFromSc
...
+      sc_json & children);
+
+  void UpdateChildArcs(
+      ScStructureElementInfo::StructureElementInfoList const & arcs,
+      bool isStruct,
+      sc_json & fullChild,
+      String const & direction);
+
+  //! Get json of arc
+  void GetChild(ScStructureElementInfo * arcInfo, String const & direction, bool isStruct, sc_json & child);
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.h:93]:
```suggestion
  void ParseScnJsonChild(ScStructureElementInfo * arcInfo, String const & direction, bool isStruct, sc_json & child);
```

**Review by NikitaZotov — APPROVED** [review]

**Review by NikitaZotov — CHANGES_REQUESTED** [review]
Fix style

File: docs/sc-kpm/sc_ui_scn_json_translator.tex [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+\scniselement{линейный язык}
+\scnsubset{JSON}
+\begin{scnreltolist}{разбиение}
+    \scnitem{scn-json-node}
+    \begin{scnindent}
+        \begin{scnreltolist}{разбиение}
+            \scnitem{scn-json-base-info}
+            \scnitem{scn-json-children}
+            \begin{scnindent}
+                \begin{scnreltolist}{разбиение}
```
Comment (NikitaZotov) [inline docs/sc-kpm/sc_ui_scn_json_translator.tex:16]:
```suggestion
                \begin{scnrelfromlist}{разбиение}
```

File: docs/sc-kpm/sc_ui_scn_json_translator.tex [hunk 28]
```diff
@@ -0,0 +1,202 @@
...
+                    \scnitem{scn-json-linked-nodes}
+                    \begin{scnindent}
+                        \begin{scnreltolist}{разбиение}
+                            \scnitem{scn-json-node}
+                        \end{scnreltolist}
+                    \end{scnindent}
+                    \scnitem{scn-json-arcs}
+                    \begin{scnindent}
+                        \begin{scnreltolist}{разбиение}
+                            \scnitem{scn-json-arc}
```
Comment (NikitaZotov) [inline docs/sc-kpm/sc_ui_scn_json_translator.tex:26]:
```suggestion
                            \scnitem{scn-json-коннектор}
```

File: docs/sc-kpm/sc_ui_scn_json_translator.tex [hunk 29]
```diff
@@ -0,0 +1,202 @@
...
+                        \end{scnreltolist}
+                    \end{scnindent}
+                    \scnitem{scn-json-modifiers}
```
Comment (NikitaZotov) [inline docs/sc-kpm/sc_ui_scn_json_translator.tex:29]:
```suggestion
                    \scnitem{scn-json-модификатор}
```

File: docs/sc-kpm/sc_ui_scn_json_translator.tex [hunk 30]
```diff
@@ -0,0 +1,202 @@
...
+                    \begin{scnindent}
+                        \begin{scnreltolist}{разбиение}
```
Comment (NikitaZotov) [inline docs/sc-kpm/sc_ui_scn_json_translator.tex:31]: Включение

File: docs/sc-kpm/sc_ui_scn_json_translator.tex [hunk 31]
```diff
@@ -0,0 +1,209 @@
...
+                        \end{scnreltolist}
+                    \end{scnindent}
+                    \scnitem{scn-json-модификаторы}
+                    \begin{scnindent}
+                        \begin{scnrelfromlist}{включение}
+                            \scnitem{scn-json-модификатор}
+                            \begin{scnindent}
+                                \begin{scnrelfromlist}{разбиение}
+                                    \scnitem{scn-json-спецификация sc-элементов}
+                                    \scnitem{scn-json-спецификация sc-коннекторов для заданного scn-json-модификатор}
```
Comment (NikitaZotov) [inline docs/sc-kpm/sc_ui_scn_json_translator.tex:36]:
```suggestion
                                    \scnitem{scn-json-спецификация sc-коннекторов для заданного scn-json-модификатора}
```

File: docs/sc-kpm/sc_ui_scn_json_translator.tex [hunk 32]
```diff
@@ -0,0 +1,209 @@
...
+\scnrelfrom{пример входной конструкции}{\scnfileimage[15em]{../images/kpm/scn_json_translator_input2.png}}
+\begin{scnrelfromlist}{аргументы агента}
+    \scnitem{question\_instance}
+    \begin{scnindent}
+        \scntext{пояснение}{действие, которое необходимо протранслировать, у которого есть ответ связанный отношением nrel\_answer}
+        \scntext{отношение аргумента}{исходная sc-конструкция'}
+    \end{scnindent}
+    \scnitem{format}
+    \begin{scnindent}
+        \scntext{пояснение}{формат в которой необходимо протранслировать, в данном случае формат scn (json)}
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply docs/sc-kpm/sc_ui_scn_json_translator.tex:183]: везде пересмотри

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 33]
```diff
@@ -4,337 +4,657 @@
...
+    sc_helper_resolve_system_identifier(s_default_ctx, formatStr.data(), &format);
+    if (sc_helper_check_arc(s_default_ctx, elInfo->addr, format, sc_type_arc_common | sc_type_const) == SC_TRUE)
+    {
+      contentType = formatStr.data();
+      break;
+    }
   }
-  else
+  // if the link has no format, then we get the content of the link
+  if (contentType.is_null())
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:246]: Divide this condition into more atomic conditions

File: sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp [hunk 34]
```diff
@@ -4,337 +4,657 @@
...
-    uiSCnSentenceNode * sentence = new uiSCnSentenceNode(mScElementsInfo[*it]);
-    mRootSentences.push_back(sentence);
+    linkedNode = arcInfo->sourceInfo;
   }
 
-  // build sentence trees
-  uiSCnSentenceNode::tSentenceNodeList::iterator itSentence, itSentenceEnd = mRootSentences.end();
-  for (itSentence = mRootSentences.begin(); itSentence != itSentenceEnd; ++itSentence)
-    (*itSentence)->buildTree();
+  sc_json arc;
```
Comment (NikitaZotov) [inline sc-kpm/sc-ui/translators/uiSc2SCnJsonTranslator.cpp:544]:
```suggestion
  ScJson arc;
```

File: sc-kpm/tests/translators/ui-scn-json-translator.cpp [hunk 35]
```diff
@@ -0,0 +1,286 @@
...
+  int const WAIT_TIME = 2;
+  int waitTime = 0;
+  ScAddr translation = findTranslation(context, answerAddr);
+  while (!translation.IsValid())
+  {
+    sleep(1);
+    waitTime += 1;
+    if (waitTime > WAIT_TIME)
+    {
+      return {};
```
Comment (NikitaZotov) [inline sc-kpm/tests/translators/ui-scn-json-translator.cpp:45]:
```suggestion
      return ScAddr::Empty;
```

Comment (NikitaZotov) [reply sc-kpm/tests/translators/ui-scn-json-translator.cpp:45]: Recheck it everywhere

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: docs/sc-kpm/sc_ui_scn_json_translator.tex [hunk 36]
```diff
@@ -0,0 +1,208 @@
...
+    \scnitem{класс sc-элемента\scnsupergroupsign}
+    \scnitem{системный идентификатор* $\cup$ основной идентификатор*}
+    \scnitem{modifierArcs}
+\end{scnrelfromlist}
+\scnrelfrom{пример}{Пример модификтора дуг родительского элемента на SCn-JSON-коде}
+\begin{scnindent}
+    \scneqimage[15em]{../images/kpm/modifier_example.png}
+\end{scnindent}
+
+\scnheader{scn-json-modifier-arcs}
```
Comment (NikitaZotov) [inline docs/sc-kpm/sc_ui_scn_json_translator.tex:146]:
```suggestion
\scnheader{scn-json-спецификация sc-коннекторов, следующих за модификатором}
```

File: docs/sc-kpm/sc_ui_scn_json_translator.tex [hunk 37]
```diff
@@ -0,0 +1,208 @@
...
+        \scntext{отношение аргумента}{исходная sc-конструкция'}
+    \end{scnindent}
+    \scnitem{format}
+    \begin{scnindent}
+        \scntext{пояснение}{формат в которой необходимо протранслировать, в данном случае формат SCn-JSON-код}
+        \scntext{отношение аргумента}{выходной формат'}
+    \end{scnindent}
+    \scnitem{lang}
+    \begin{scnindent}
+        \scntext{пояснение}{язык содержимого sc-ссылок и идентификаторов}
```
Comment (NikitaZotov) [inline docs/sc-kpm/sc_ui_scn_json_translator.tex:187]:
```suggestion
        \scntext{пояснение}{язык содержимого файлов ostis-системы}
```

File: sc-kpm/tests/translators/test-structures/command_init_checks/init_ui_translator.scs [hunk 38]
```diff
@@ -0,0 +1,20 @@
...
+system_element -> @edge1;;
+system_element -> (ui_rrel_source_sc_construction-> @edge1);;
+@edge2 = (trans_cmd_addr -> lang_ru);;
+system_element -> @edge2;;
+system_element -> (ui_rrel_user_lang-> @edge2);;
+@edge3 = (trans_cmd_addr -> format_scn_json);;
+system_element -> @edge3;;
+system_element -> (ui_rrel_output_format-> @edge3);;
+system_element -> (trans_cmd_addr <- ui_command_translate_from_sc);;
+question_addr => nrel_answer: answer_addr;;
```
Comment (NikitaZotov) [inline sc-kpm/tests/translators/test-structures/command_init_checks/init_ui_translator.scs:20]: newline

File: sc-kpm/tests/translators/test-structures/command_init_checks/command_init_no_format.scs [hunk 39]
```diff
@@ -0,0 +1,16 @@
...
+ui_rrel_source_sc_construction;
+ui_rrel_user_lang;
+ui_rrel_output_format;;
+@edge1 = (trans_cmd_addr -> answer_addr);;
+system_element -> @edge1;;
+system_element -> (ui_rrel_source_sc_construction-> @edge1);;
+@edge2 = (trans_cmd_addr -> lang_ru);;
+system_element -> @edge2;;
+system_element -> (ui_rrel_user_lang-> @edge2);;
+system_element -> (trans_cmd_addr <- ui_command_translate_from_sc);;
```
Comment (NikitaZotov) [inline sc-kpm/tests/translators/test-structures/command_init_checks/command_init_no_format.scs:15]: newline

**Review by NikitaZotov — COMMENTED** [review]
I checked code by valgrind tool. There are no memory leaks.

**Comment by NikitaZotov** [comment]
Clean history and merge it!

**Review by NikitaZotov — APPROVED** [review]

**Pull request merged**